File: js/src/prompt-cache/parameters-cache.ts
import { RemoteEvalParameters } from "../logger";
import { LRUCache } from "../lru-cache";
import { DiskCache } from "./disk-cache";

interface ParametersKey {
  slug?: string;
  version?: string;
  projectId?: string;
  projectName?: string;
  id?: string;
}

function createCacheKey(key: ParametersKey, namespace?: string): string {
  let cacheKey: string;
  if (key.id) {
    cacheKey = `parameters:id:${key.id}`;
  } else {
    const prefix = key.projectId ?? key.projectName;
    if (!prefix) {
      throw new Error("Either projectId or projectName must be provided");
    }
    if (!key.slug) {
      throw new Error("Slug must be provided when not using ID");
    }
    cacheKey = `parameters:${prefix}:${key.slug}:${key.version ?? "latest"}`;
  }
  return namespace === undefined
    ? cacheKey
    : `${namespace.length}:${namespace}:${cacheKey}`;
}

export type ParametersMemoryCacheEntry = {
  value: RemoteEvalParameters;
  resolvedOrgIdentity?: string;
};

export type ParametersDiskCacheEntry = {
  value: ReturnType<RemoteEvalParameters["_internalSerializeForCache"]>;
  resolvedOrgIdentity?: string;
};

export class ParametersCache {
  private readonly memoryCache?: LRUCache<string, ParametersMemoryCacheEntry>;
  private readonly diskCache?: DiskCache<ParametersDiskCacheEntry>;
  private readonly namespace?: string;
  private readonly expectedResolvedOrgIdentity?: string;

  constructor(options: {
    memoryCache?: LRUCache<string, ParametersMemoryCacheEntry>;
    diskCache?: DiskCache<ParametersDiskCacheEntry>;
    namespace?: string;
    expectedResolvedOrgIdentity?: string;
  }) {
    this.memoryCache = options.memoryCache;
    this.diskCache = options.diskCache;
    this.namespace = options.namespace;
    this.expectedResolvedOrgIdentity = options.expectedResolvedOrgIdentity;
  }

  withNamespace(
    namespace: string,
    expectedResolvedOrgIdentity?: string,
  ): ParametersCache {
    return new ParametersCache({
      memoryCache: this.memoryCache,
      diskCache: this.diskCache,
      namespace,
      expectedResolvedOrgIdentity,
    });
  }

  async get(key: ParametersKey): Promise<RemoteEvalParameters | undefined> {
    const cacheKey = createCacheKey(key, this.namespace);
    if (this.memoryCache) {
      const memoryEntry = this.memoryCache.get(cacheKey);
      if (
        memoryEntry !== undefined &&
        (this.expectedResolvedOrgIdentity === undefined ||
          memoryEntry.resolvedOrgIdentity === this.expectedResolvedOrgIdentity)
      ) {
        return memoryEntry.value;
      }
    }

    if (this.diskCache) {
      const diskEntry = await this.diskCache.get(cacheKey);
      if (
        !diskEntry ||
        (this.expectedResolvedOrgIdentity !== undefined &&
          diskEntry.resolvedOrgIdentity !== this.expectedResolvedOrgIdentity)
      ) {
        return undefined;
      }
      const diskParameters = new RemoteEvalParameters(diskEntry.value.metadata);
      this.memoryCache?.set(cacheKey, {
        value: diskParameters,
        resolvedOrgIdentity: diskEntry.resolvedOrgIdentity,
      });
      return diskParameters;
    }

    return undefined;
  }

  async set(key: ParametersKey, value: RemoteEvalParameters): Promise<void> {
    const cacheKey = createCacheKey(key, this.namespace);
    const memoryEntry = {
      value,
      resolvedOrgIdentity: this.expectedResolvedOrgIdentity,
    };
    this.memoryCache?.set(cacheKey, memoryEntry);
    if (this.diskCache) {
      await this.diskCache.set(cacheKey, {
        value: value._internalSerializeForCache(),
        resolvedOrgIdentity: this.expectedResolvedOrgIdentity,
      });
    }
  }
}

File: py/src/braintrust/prompt_cache/parameters_cache.py
from __future__ import annotations

from typing import Any

from braintrust.prompt_cache import disk_cache, lru_cache


def _create_cache_key(
    project_id: str | None,
    project_name: str | None,
    slug: str | None,
    version: str = "latest",
    id: str | None = None,
) -> str:
    if id:
        return f"parameters:id:{id}"

    prefix = project_id or project_name
    if not prefix:
        raise ValueError("Either project_id or project_name must be provided")
    if not slug:
        raise ValueError("Slug must be provided when not using ID")
    return f"parameters:{prefix}:{slug}:{version}"


class ParametersCache:
    def __init__(
        self,
        memory_cache: lru_cache.LRUCache[str, dict[str, Any]],
        disk_cache: disk_cache.DiskCache[dict[str, Any]] | None = None,
    ):
        self.memory_cache = memory_cache
        self.disk_cache = disk_cache

    def get(
        self,
        slug: str | None = None,
        version: str = "latest",
        project_id: str | None = None,
        project_name: str | None = None,
        id: str | None = None,
    ) -> dict[str, Any]:
        cache_key = _create_cache_key(project_id, project_name, slug, version, id)

        try:
            return self.memory_cache.get(cache_key)
        except KeyError:
            pass

        if self.disk_cache:
            params = self.disk_cache.get(cache_key)
            self.memory_cache.set(cache_key, params)
            return params

        raise KeyError(f"Parameters not found in cache: {cache_key}")

    def set(
        self,
        value: dict[str, Any],
        slug: str | None = None,
        version: str = "latest",
        project_id: str | None = None,
        project_name: str | None = None,
        id: str | None = None,
    ) -> None:
        cache_key = _create_cache_key(project_id, project_name, slug, version, id)

        self.memory_cache.set(cache_key, value)
        if self.disk_cache:
            self.disk_cache.set(cache_key, value)
